File: ueb4/scala-shared/shared/src/main/scala/Messages.scala
import io.circe._
import io.circe.generic.semiauto._

case class TimestampedMessage(
  timestamp: Long,
  payload: Message
)

sealed trait Message
case object Ping extends Message
case object Pong extends Message
case class Asteroid(common : CommonState, size : ASize, color : AColor) extends Message 
case class Spaceship(common : CommonState) extends Message 
case class Laser(common : CommonState, shooter : Int) extends Message
case class Cmd(sId : Int, cmd : Command) extends Message
case class ClientId(clientId : Int) extends Message
case class Destroy(destroy : Int) extends Message
case class SetLifes(sId : Int, lifes : Int) extends Message

case class CommonState(ident : Int, pos : (Double, Double), velo : (Double, Double), acc : Double, omega : Double, phi : Double)  

sealed trait AColor 
case object Brown extends AColor
case object Gray extends AColor

sealed trait ASize
case object Tiny extends ASize
case object Small extends ASize
case object Medium extends ASize
case object Big extends ASize

sealed trait Command 
case object F extends Command
case object B extends Command
case object L extends Command
case object R extends Command
case object S extends Command

///////////////////////////
// JSON De/serialization //
///////////////////////////

sealed trait JSONConfig {
  import io.circe.Encoder
  import shapeless.{ Generic, HNil }

  implicit def encodeCaseObject[A <: Product](implicit
    gen: Generic.Aux[A, HNil]
  ): Encoder[A] = Encoder.instance(_ => Json.arr())
}

object TimestampedMessage extends JSONConfig {
  def apply(payload: Message): TimestampedMessage = TimestampedMessage(System.currentTimeMillis(),payload)

  implicit val encodeTimestampedMessage: Encoder[TimestampedMessage] =
    deriveEncoder[TimestampedMessage]

  implicit val decodeTimestampedMessage: Decoder[TimestampedMessage] =
    deriveDecoder[TimestampedMessage]
}

object Message extends JSONConfig {
  implicit val encodeMessage: Encoder[Message] =
    deriveEncoder[Message]

  implicit val decodeMessage: Decoder[Message] =
    deriveDecoder[Message]
}

object CommonState extends JSONConfig {
  implicit val encodeMessage: Encoder[CommonState] =
    deriveEncoder[CommonState]

  implicit val decodeMessage: Decoder[CommonState] =
    deriveDecoder[CommonState]
}

object AColor extends JSONConfig {
  implicit val encodeMessage: Encoder[AColor] =
    deriveEncoder[AColor]

  implicit val decodeMessage: Decoder[AColor] =
    deriveDecoder[AColor]
}

object ASize extends JSONConfig {
  implicit val encodeMessage: Encoder[ASize] =
    deriveEncoder[ASize]

  implicit val decodeMessage: Decoder[ASize] =
    deriveDecoder[ASize]
}

object Command extends JSONConfig {
  implicit val encodeMessage: Encoder[Command] =
    deriveEncoder[Command]

  implicit val decodeMessage: Decoder[Command] =
    deriveDecoder[Command]
}

File: ueb4/haskell-server/Messages.hs
{-# LANGUAGE DeriveGeneric #-}

module Messages (
  Message (..),
  CommonState (..),
  AColor (..),
  ASize (..),
  Command (..),
  TimestampedMessage (..),
  currentTimeMillis,
  timestamped,
  ServerMessage (..)
) where

import GHC.Generics
import Data.Aeson
import Data.Aeson.Types
import Data.Time.Clock.POSIX (getPOSIXTime)
import Control.Concurrent.MVar
import Hakka.Actor (ActorRef)

data Message = Ping | Pong | Init | Update |   
  CreateLaser Int (Double, Double) (Double, Double) Double | 
  Cmd { sId :: Int, cmd :: Command } | 
  ClientId { clientId :: Int } |
  Asteroid { 
    common :: CommonState,
    size :: ASize,
    color :: AColor } | 
  Spaceship { common :: CommonState } |
  Laser { common :: CommonState, shooter :: Int } deriving (Generic, Show)

data Command = F | B | L | R | S deriving (Generic, Show) 

data CommonState = CommonState { 
  ident :: Int,
  pos :: (Double, Double),
  velo :: (Double, Double), 
  acc :: Double, 
  omega :: Double,
  phi :: Double
} deriving (Generic, Show) 

data ASize = Tiny | Small | Medium | Big deriving (Generic, Show) 
data AColor = Brown | Gray deriving (Generic, Show) 

currentTimeMillis :: IO Integer
currentTimeMillis = (round . (1000 *)) <$> getPOSIXTime

timestamped :: Message -> IO TimestampedMessage
timestamped msg = currentTimeMillis >>= \time -> return $ TimestampedMessage time msg

data TimestampedMessage = TimestampedMessage {
  timestamp :: Integer,
  payload :: Message
} deriving (Generic,Show)

data ServerMessage = 
    Connect {
      socketOut :: (TimestampedMessage -> IO ()),
      connectionActor :: MVar (ActorRef ServerMessage)
    }
  | Disconnect Int
  | Msg TimestampedMessage
  | Ok

instance Show ServerMessage where
  show (Connect _ _)  = "Connect <...>"
  show (Disconnect _) = "Disconnect"
  show (Msg x) = "Msg " ++ show x
  show Ok = "Ok"

----------------------
-- JSON En/Decoding --

jsonOptions :: Options
jsonOptions = defaultOptions {
  sumEncoding = ObjectWithSingleField,
  allNullaryToStringTag = False
}

instance ToJSON Message where
  toEncoding = genericToEncoding jsonOptions

instance FromJSON Message where
  parseJSON = genericParseJSON jsonOptions

instance ToJSON TimestampedMessage where
  toEncoding = genericToEncoding jsonOptions

instance FromJSON TimestampedMessage where
  parseJSON = genericParseJSON jsonOptions

instance ToJSON CommonState where
  toEncoding = genericToEncoding jsonOptions

instance FromJSON CommonState where
  parseJSON = genericParseJSON jsonOptions

instance ToJSON ASize where
  toEncoding = genericToEncoding jsonOptions

instance FromJSON ASize where
  parseJSON = genericParseJSON jsonOptions

instance ToJSON AColor where
  toEncoding = genericToEncoding jsonOptions

instance FromJSON AColor where
  parseJSON = genericParseJSON jsonOptions

instance ToJSON Command where
  toEncoding = genericToEncoding jsonOptions

instance FromJSON Command where
  parseJSON = genericParseJSON jsonOptions
